runtime/compiler/optimizer/abstractinterpreter/AbsInterpreter.cpp

```diff
@@ -0,0 +1,3072 @@
+                                                      true);
+         if(yesNoMaybe == TR_yes) //Instanceof must be true;
+            {
+            state->push(createIntConst(1));
+            return;
+            }
+         else if (yesNoMaybe == TR_no) //Instanceof must be false;
+            {
+            state->push(createIntConst(0));
+            return;
+            }
+         }
+      }
+
+   state->push(createIntRange(0, 1));
+   return;
+   }
+
+void TR::AbsBlockInterpreter::checkcast()
+   {
+   TR::AbsStackMachineState* state = getState();
+
+   TR::AbsValue *objRef = state->pop();
+   TR_ASSERT_FATAL(objRef->getDataType() == TR::Address, "Unexpected type");
+
+   int32_t cpIndex = _bci.next2Bytes();
+   TR_OpaqueClassBlock* castClass = _callerMethod->getClassFromConstantPool(comp(), cpIndex);
+
+   //adding to method summary
+   if (objRef->isParameter() && !(!_callerMethod->isStatic() && objRef->getParameterPosition() == 0) && lastTimeThrough())
+      {
+      TR::PotentialOptimizationVPPredicate* p1 =
+         new (region()) TR::PotentialOptimizationVPPredicate(TR::VPNullObject::create(vp()), _bci.currentByteCodeIndex(), TR::PotentialOptimizationPredicate::Kind::CheckCastFolding, vp());
+
+      _inliningMethodSummary->addPotentialOptimizationByArgument(p1, objRef->getParameterPosition());
+
+      if (castClass)
+         {
+         TR::PotentialOptimizationVPPredicate* p2 =
+            new (region()) TR::PotentialOptimizationVPPredicate(TR::VPResolvedClass::create(vp(), castClass), _bci.currentByteCodeIndex(), TR::PotentialOptimizationPredicate::Kind::CheckCastFolding, vp());
+
+         _inliningMethodSummary->addPotentialOptimizationByArgument(p2, objRef->getParameterPosition());
+         }
+      }
+
+   if (isNullObject(objRef)) //Check cast null object, always succeed
+      {
+      state->push(objRef);
+      return;
+      }
+
+   if (isObject(objRef) && isNonNullObject(objRef))
+      {
+      if (castClass && static_cast<TR::AbsVPValue*>(objRef)->getConstraint()->getClass())
+         {
+         TR_YesNoMaybe yesNoMaybe = comp()->fe()->isInstanceOf(
+                                          static_cast<TR::AbsVPValue*>(objRef)->getConstraint()->getClass(),
+                                          castClass,
+                                          static_cast<TR::AbsVPValue*>(objRef)->getConstraint()->isFixedClass(),
+                                          true);
+         if (yesNoMaybe == TR_yes)
+            {
+            if (castClass == static_cast<TR::AbsVPValue*>(objRef)->getConstraint()->getClass()) //cast into the same type, no change
+               {
+               state->push(objRef);
+               return;
+               }
+            else //cast into a different type
+               {
+               state->push(createObject(castClass, TR_yes));
+               return;
+               }
+            }
+         }
+      }
+
+   state->push(createTopObject());
```

Comment:
I added another if statement to check if we have `castClass`, if yes, the result that will be push into the stack is `createObject(castClass)` (or formerly `createObject(castClass, TR_maybe`). Otherwise, the returned object of `createTopObject()` will be push into the stack.

I think even we know the `objRef` is non-null, we should do the checkcast first, and if the checkcast is NOT guaranteed to succeed, it will still fall into check if we have `castClass` and do `createObject(castClass)`. So I didn't add another if statement for the situation that `objRef` is non-null. OR, you mean under that situation we should do `createNonNullObject(castClass)` (which is formerly `createObject(castClass, TR_yes)`)?

Added parameter position from `objRef` to both objects created by `createObject(castClass)` and `createTopObject()`.